_studio/shared/src/mfx_umc_mjpeg_vpp.cpp

```diff
@@ -180,6 +180,19 @@ mfxStatus mfx_UMC_FrameAllocator_D3D_Converter::StartPreparingToOutput(mfxFrameS
 
     mfxStatus sts = MFX_ERR_NONE;
     bool isD3DToSys = false;
+    mfxVideoParam  localParams;
+    const mfxVideoParam *par = parArg;
```

Comment:
This is ugly construction. Since we are creating new variable of type ```mfxVideoParam``` let's change function signature to accept ```mfxVideoParam``` by value (i.e. to perform copying on function call) and then change local copy of ```mfxVideoParam``` in place

Reply:
Well, it's your decision. I though that rotation is rare case, so we may want to optimize and do copy only when this is necessary. sizeof(mfxVideoParam) is 208 Bytes, so the copy is not that free.

Reply:
You've already wasted such amount of bytes on tmp variable, so savings related only to copy. As for me it not worth it. Let's wait for some other opinion

Reply:
Stack allocation is free as far as it's not touched, and `mfxVideoParam` doesn't have a constructor.

Reply:
Construction like
```
localParams = *parArg; 
//...
par = &localParams; 
```
is acceptable while a function is short and this piece is at the beginning of function. Otherwise it might be surprising to reveal at point where `par` is used that its address was changed somewhere.

I'd rather use
```
mfxVideoParam const& par //as method arg
//...
mfxVideoParam  vp = par; //local copy of FrameInfo
// ... using of vp.XXX instead par->XXX
```
until we found that we have perf. drop because of copying.

BTW, that's why I argue against `mfxVideoParam` as one universal argument everywhere. It looks like we really don't need the whole `mfxVideoParam` here, just frame info. The only place where it is needed is `DoHwJpegCcFw` for `VideoVppJpegD3D9` initialization. The latter is odd thing for me itself - why we can't create & initialize it explicitly, at decoder's `Init` method? At `Init` we have no any conserns about how many bytes is `sizeof(mfxVideoParam)` and is it worth to avoid copying or not.

Reply:
> I'd rather use
> mfxVideoParam const& par //as method arg
> //...

Sorry, I can't see why it is better than just passing `mfxVideoParam` by value, as Miroslav suggested above.

> until we found that we have perf. drop because of copying.

I bet we don't, because this is called once for a frame. But a problem with such approach is that it creates tons of very small bottlenecks, that is real pain if someone need performance improvement after all.

Reply:
@Alexandr-Konovalov,
> why it is better than just passing mfxVideoParam by value

because it (passing structs by value) is
1. quite rare code pattern thereof looks like a typo and pushes reader to go deeper into the details to find out why it is done in that way
2. implicit as against `mfxVideoParam  vp = par;` where we show our intention clearly

> it creates tons of very small bottlenecks
>real pain if someone need performance improvement

IMO, clean design is the first until performance analyzer says that we need sacrifice it.

Let's refactor code in a way when `mfxVideoParam` is not needed anymore at that function. It makes design clearer and allow us to avoid these (very potential) bottlenecks.

Reply:
Yes, I also vote for refactoring as best solution.

Reply:
I still don't understand what's unclear with creation of duplicates only when we need them. OK, let's do it your way.

Getting rid of `DoHwJpegCc* ` etc seems right thing.